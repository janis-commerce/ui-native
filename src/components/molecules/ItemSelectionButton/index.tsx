import React from 'react';
import RadioButton from 'atoms/RadioButton';
import CheckBox from 'atoms/CheckBox';
import {StyleSheet, View} from 'react-native';
import Text from 'atoms/Text';
import {palette} from 'theme/palette';
import {verticalScale} from 'scale';

export interface ItemSelectionButtonProps {
	name: string;
	selected?: boolean;
	radioButton?: boolean;
	leftSelection?: boolean;
	rightSelection?: boolean;
	onSelection?: () => {};
}

const styles = StyleSheet.create({
	container: {
		display: 'flex',
		height: verticalScale(48),
		borderRadius: verticalScale(4),
	},
	checkBoxContainer: {
		justifyContent: 'space-between',
	},
	text: {
		fontSize: 16,
		fontWeight: '400',
		color: palette.black.secondary,
		textTransform: 'capitalize',
	},
});

const ItemSelectionButton: React.FC<ItemSelectionButtonProps> = ({
	name,
	selected = false,
	radioButton = false,
	leftSelection = false,
	rightSelection = true,
	onSelection = () => {},
}) => {
	if (!name) {
		return null;
	}

	const checkPosition = rightSelection ? 'right' : 'left';

	const renderSelectionComponent = () => {
		if (!radioButton) {
			return <CheckBox onPress={onSelection} checked={selected} customSize={24} />;
		}

		return (
			<RadioButton
				onPress={onSelection}
				selected={selected}
				checkPosition={checkPosition}
				checkSize="md">
				<Text style={styles.text}>{name}</Text>
			</RadioButton>
		);
	};

	return (
		<View style={styles.container}>
			{leftSelection && !rightSelection && renderSelectionComponent()}
			{!radioButton && <Text style={styles.text}>{name}</Text>}
			{rightSelection && !leftSelection && renderSelectionComponent()}
		</View>
	);
};

export default ItemSelectionButton;
