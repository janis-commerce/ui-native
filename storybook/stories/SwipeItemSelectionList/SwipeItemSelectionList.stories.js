import React from 'react';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {StyleSheet, Text, View} from 'react-native';
import SwipeItemSelectionList from 'organisms/SwipeItemSelectionList';

export default {
	title: 'Components/SwipeItemSelectionList',
	argTypes: {
		snapPoints: {
			options: ['5_50_100_percentage', '120_240_300_px'],
			mapping: {
				'5_50_100_percentage': ['5%', '50%', '100%'],
				'120_240_300_px': [120, 240, 300],
			},
			control: {type: 'radio'},
		},
		snapPosition: {
			options: [0, 1, 2],
			control: {type: 'radio'},
		},
		multiselect: {
			options: [true, false],
		},
		renderHeader: {
			options: [true, false],
		},
	},
};

const styles = StyleSheet.create({
	screenStyle: {
		flex: 1,
		height: 400,
	},
	headerContainer: {
		justifyContent: 'center',
		alignItems: 'center',
	},
	headerTitle: {
		fontWeight: 'bold',
		fontSize: 25,
	},
});

const options = Array.from({length: 30}, (_, i) => ({
	id: (i + 1).toString(),
	name: `Warehouse ${(i + 1).toString()}`,
}));

export const WithSwipeUpViewComponent = (props) => {
	const {renderHeaderComponent, headerText, showActions, multiselect} = props;

	const actions = [{value: 'Confirmar'}];

	const renderHeader = () => {
		if (!renderHeaderComponent) {
			return null;
		}

		return (
			<View style={styles.headerContainer}>
				<Text style={styles.headerTitle}>{headerText}</Text>
			</View>
		);
	};

	return (
		<GestureHandlerRootView style={styles.screenStyle}>
			<SwipeItemSelectionList
				data={options}
				renderHeader={renderHeader}
				multiselect={multiselect}
				{...(showActions && actions)}
				{...props}
			/>
		</GestureHandlerRootView>
	);
};

WithSwipeUpViewComponent.args = {
	snapPoints: ['25%', '50%', '100%'],
	snapPosition: 0,
	renderHeaderComponent: false,
	headerText: 'Header',
	showActions: false,
	multiselect: false,
};
