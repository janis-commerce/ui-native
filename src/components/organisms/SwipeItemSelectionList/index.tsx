import React, {useState} from 'react';
import SwipeList, {SwipeListProps} from 'molecules/SwipeList';
import ItemSelectionButton from 'molecules/ItemSelectionButton';

interface SwipeItemSelectionListProps extends Partial<SwipeListProps> {
	data: any[];
	multiselect?: boolean;
	leftSelection?: boolean;
	rightSelection?: boolean;
	onSelection?: (id: string) => {};
}

const SwipeItemSelectionList: React.FC<SwipeItemSelectionListProps> = ({
	data,
	multiselect = false,
	leftSelection = false,
	rightSelection = false,
	onSelection = () => {},
	...props
}) => {
	const [selectedElementId, setSelectedElementId] = useState<string>('');
	const [selectedElementsIds, setSelectedElementsIds] = useState<string[]>([]);

	if (!data || !data.length) {
		return null;
	}

	const checkIfElementIsSelected = (id: string, isElementSelected: boolean) => {
		if (isElementSelected) {
			return multiselect
				? setSelectedElementsIds((prevState) => prevState.filter((includedId) => includedId !== id))
				: setSelectedElementId('');
		}

		return multiselect
			? setSelectedElementsIds((prevState) => [...prevState, id])
			: setSelectedElementId(id);
	};

	const renderItem = (element: {id: string; name: string}) => {
		const {id, name} = element;

		const isElementSelected = multiselect
			? selectedElementsIds.includes(id)
			: selectedElementId === id;

		const onSelectElement = () => {
			checkIfElementIsSelected(id, isElementSelected);
			return onSelection(id);
		};

		return (
			<ItemSelectionButton
				radioButton={multiselect}
				leftSelection={leftSelection}
				rightSelection={rightSelection}
				selected={isElementSelected}
				onSelection={onSelectElement as () => {}}
				name={name}
			/>
		);
	};

	return <SwipeList {...props}>{data.map((element) => renderItem(element))}</SwipeList>;
};

export default SwipeItemSelectionList;
