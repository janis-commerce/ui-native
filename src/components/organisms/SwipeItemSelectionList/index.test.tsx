import 'react-native';
import React from 'react';
import {create} from 'react-test-renderer';
import SwipeItemSelectionList from './index';
import ItemSelectionButton from 'molecules/ItemSelectionButton';
import {act, fireEvent} from '@testing-library/react-native';

describe('SwipeItemSelectionList component', () => {
	describe('does not render', () => {
		it('as no data is passed', () => {
			const {toJSON} = create(<SwipeItemSelectionList data={[]} />);

			expect(toJSON()).toBeNull();
		});
	});

	describe('renders correctly, select and unselect it', () => {
		it('with checkbox, selects and unselects it', async () => {
			const {toJSON, root} = create(<SwipeItemSelectionList data={[{id: '1', name: '1'}]} />);

			const ItemSelectionButtonComp = root.findByType(ItemSelectionButton);

			await act(async () => fireEvent(ItemSelectionButtonComp, 'onSelection'));
			await act(async () => fireEvent(ItemSelectionButtonComp, 'onSelection'));

			expect(toJSON()).toMatchSnapshot();
		});

		it('with radioButton, selects and unselect it', async () => {
			const {toJSON, root} = create(
				<SwipeItemSelectionList data={[{id: '1', name: '1'}]} multiselect />
			);

			const ItemSelectionButtonComp = root.findByType(ItemSelectionButton);
			await act(async () => fireEvent(ItemSelectionButtonComp, 'onSelection'));
			await act(async () => fireEvent(ItemSelectionButtonComp, 'onSelection'));

			expect(toJSON()).toMatchSnapshot();
		});
	});
});
