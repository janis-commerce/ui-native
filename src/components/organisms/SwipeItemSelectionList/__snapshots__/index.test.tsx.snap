// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeItemSelectionList component renders correctly, select and unselect it with checkbox, selects and unselects it 1`] = `
<View
  footerComponent={[Function]}
  index={0}
  snapPoints={
    Array [
      "100%",
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "justifyContent": "center",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "auto",
              "justifyContent": "space-between",
              "marginVertical": 27,
              "opacity": 1,
              "paddingHorizontal": 33.5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Roboto",
                  "fontSize": 27,
                },
                Object {
                  "color": "#2F2F2F",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            1
          </Text>
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderRadius": 12.5,
                "height": 50,
                "opacity": 1,
                "width": 50,
              }
            }
          >
            <View
              style={
                Object {
                  "borderColor": "#939598",
                  "borderRadius": 6,
                  "borderWidth": 1.5,
                  "height": 50,
                  "width": 50,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "borderBottomColor": "#D5D7DB",
              "borderBottomWidth": 2,
              "width": "90%",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`SwipeItemSelectionList component renders correctly, select and unselect it with radioButton, selects and unselect it 1`] = `
<View
  footerComponent={[Function]}
  index={0}
  snapPoints={
    Array [
      "100%",
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "justifyContent": "center",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "auto",
              "justifyContent": "space-between",
              "marginVertical": 27,
              "opacity": 1,
              "paddingHorizontal": 33.5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Roboto",
                  "fontSize": 27,
                },
                Object {
                  "color": "#2F2F2F",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            1
          </Text>
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderRadius": 12.5,
                "height": 50,
                "opacity": 1,
                "width": 50,
              }
            }
          >
            <View
              style={
                Object {
                  "borderColor": "#939598",
                  "borderRadius": 6,
                  "borderWidth": 1.5,
                  "height": 50,
                  "width": 50,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "borderBottomColor": "#D5D7DB",
              "borderBottomWidth": 2,
              "width": "90%",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
